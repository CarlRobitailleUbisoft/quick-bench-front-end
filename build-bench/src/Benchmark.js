import React from 'react';
import CodeEditor from './CodeEditor.js';
import BashOutput from 'components/BashOutput.js';
import CompileConfig from './CompileConfig.js';
import TimeChart from './TimeChart.js';
import { Button, ButtonToolbar, Row, Col, Container, Card, FormCheck, Form, ProgressBar, Nav, Tab } from 'react-bootstrap';
import { MdTimer } from "react-icons/md";
import OutputTabs from './OutputTabs.js';
import WrappableTabs from './WrappableTabs.js';
import DisplayEditor from './DisplayEditor.js';
import IncludesDisplay from './IncludesDisplay.js';

var request = require('request');
const protocolVersion = 3;

const startCode1 = `#include <cstdio>

int main() {
    puts("Hello World");
    return 0;
}
`;
const startCode2 = `#include <iostream>

int main() {
    std::cout << "Hello World\\n";
    return 0;
}
`;
const chartData = [{
    title: ["Compilation CPU Time", "Lower is faster"],
    property: ["kernelTime", "userTime"],
    name: "Time",
    more: "slower",
    less: "faster",
    xaxis: "xstacks",
    yaxis: "ystacks"
}, {
    title: "Maximum Resident Set Size (kB)",
    property: "memory",
    name: "Memory",
    more: "more",
    less: "less",
    xaxis: "bar",
    yaxis: "linear"
}, {
    title: "Number of Filesystem Inputs",
    property: "inputs",
    name: "I/O Reads",
    more: "more",
    less: "less",
    xaxis: "bar",
    yaxis: "linear"
}];
class Benchmark extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            texts: [startCode1, startCode2]
            , titles: ['cstdio', 'iostream']
            , graph: []
            , messages: ['', '']
            , sending: false
            , progress: 0
            , index: 0
            , options: Array(2).fill().map(a => ({
                compiler: "clang-9.0"
                , cppVersion: "20"
                , optim: "3"
                , lib: "gnu"
            }))
            , clean: false
            , force: false
            , benchNames: []
            , location: props.id
            , annotation: ''
            , isAnnotated: true
            , assemblyFull: false
            , chartIndex: 0
            , textsWrapped: false
            , optionsWrapped: true
            , includes: []
            , asm: []
            , pp: []
        };

        let stateFromHash = this.getStateFromHash();
        if (stateFromHash) {
            this.state.text = stateFromHash.text;
            if (stateFromHash.compiler) this.state.compiler = stateFromHash.compiler;
            if (stateFromHash.cppVersion) this.state.cppVersion = stateFromHash.cppVersion;
            if (stateFromHash.optim) this.state.optim = stateFromHash.optim;
            if (stateFromHash.lib) this.state.lib = stateFromHash.lib;
        }

        this.graph = [];
        this.url = this.props.url;
        this.maxCodeSize = this.props.maxCodeSize;
    }
    getStateFromHash() {
        if (window.location.hash) {
            let state = this.decodeHash(window.location.hash);
            window.location.hash = "";
            if (state.text) {
                return state;
            }
        }

        return false;
    }
    componentDidMount() {
        if (this.props.id) {
            this.getCode(this.props.id);
        }
        this.props.onDisplay();
    }
    componentWillReceiveProps(nextProps) {
        if (this.props.id !== nextProps.id && this.state.location !== nextProps.id) {
            this.getCode(nextProps.id);
        }
    }
    median(values) {
        if (values.length === 0) return 0;

        values.sort((a, b) => a - b);
        var half = Math.floor(values.length / 2);

        if (values.length % 2)
            return values[half];

        return (values[half - 1] + values[half]) / 2.0;
    }
    makeGraph(result, titles) {
        return result.map((r, i) => ({ times: r.times, memories: r.memories, inputs: r.inputs, title: titles[i] })).filter(r => typeof r.times !== 'undefined' && r.times.length).map((r, i) => {
            let userTime = r.times.map(t => parseFloat(t.user)).reduce((s, t) => (s + t)) / r.times.length;
            let kernelTime = r.times.map(t => parseFloat(t.kernel)).reduce((s, t) => (s + t)) / r.times.length;
            let memories = this.median(r.memories.map(n => parseFloat(n)));
            let inputs = this.median(r.inputs.map(n => parseFloat(n)));
            return {
                x: r.title,
                userTime: userTime,
                kernelTime: kernelTime,
                inputs: inputs,
                memory: memories
            };
        });
    }
    bufferMap(buffers) {
        return (buffers || []).map(s => s ? Buffer.from(s).toString() : '');
    }
    formatIncludes(includes) {
        return includes.map(s => s.split('\n').map(s => {
            let end = false;
            return [...s].map(c => {
                if (c === '.' && !end)
                    return '\t'
                end = true;
                return c;
            }).join('').replace('\t ', '');
        }).join('\n'));
    }
    getCode(id) {
        this.setState({
            sending: true,
            graph: [],
            annotation: '',
            messages: []
        });
        request.get(this.url + '/build/' + id, (err, res, body) => {
            this.setState({
                sending: false,
                clean: true,
                force: false
            });
            if (body) {
                let result = JSON.parse(body);
                if (result) {
                    if (result.result) {
                        let titles = result.tabs.map(t => t.title);
                        let options = result.tabs.map(t => ({
                            compiler: t.compiler
                            , cppVersion: t.cppVersion
                            , optim: t.optim
                            , lib: t.lib
                        }));
                        this.setState({
                            texts: result.tabs.map(t => t.code)
                            , titles: titles
                            , graph: this.makeGraph(result.result, titles)
                            , options: options
                            , location: id
                            , textsWrapped: result.tabs.every(v => v.code === result.tabs[0].code)
                            , optionsWrapped: options.every(o => JSON.stringify(o) === JSON.stringify(options[0]))
                            , includes: this.formatIncludes(result.includes)
                            , asm: result.asm
                            , pp: result.preprocessed
                        });
                    }
                    if (result.messages) {
                        this.setState({
                            messages: result.messages
                        });
                    }
                }
            }
        });
    }
    sendCode() {
        if (this.state.texts.some(t => t.length > this.maxCodeSize)) {
            this.setState({
                graph: [],
                annotation: '',
                messages: [`Your code is ${this.state.texts.length} characters long, while the maximum code size is ${this.maxCodeSize}.
If you think this limitation is stopping you in a legitimate usage of build-bench, please contact me.`]
            });
        } else {
            this.setState({
                sending: true,
                graph: [],
                annotation: '',
                messages: []
            });
            this.setState({ progress: 0 });
            let interval = setInterval(() => {
                this.setState({ progress: this.state.progress + 100 / 120 });
            }, 1000);

            var obj = {
                "tabs": this.state.texts.map((c, i) => {
                    return {
                        "code": c,
                        "title": this.state.titles[i],
                        "compiler": this.state.options[i].compiler,
                        "optim": this.state.options[i].optim,
                        "cppVersion": this.state.options[i].cppVersion,
                        "lib": this.state.options[i].lib
                        , "asm": 'att'
                        , "withPP": true
                    }
                }),
                "protocolVersion": protocolVersion,
                "force": this.state.clean && this.state.force,
            };
            request({
                url: this.url + '/build/'
                , method: "POST"
                , json: true
                , headers: {
                    "content-type": "application/json"
                }
                , body: obj
            }, (err, res, body) => {
                this.setState({
                    sending: false,
                    clean: true,
                    force: false
                });
                clearInterval(interval);
                if (body) {
                    if (body.result) {
                        let g = this.makeGraph(body.result, this.state.titles)
                        this.setState({
                            graph: g,
                            location: body.id,
                            includes: this.formatIncludes(this.bufferMap(body.includes)),
                            asm: this.bufferMap(body.asm),
                            pp: this.bufferMap(body.preprocessed)
                        });
                        this.props.onLocationChange(body.id);
                    }
                    if (body.annotation) {
                        this.setState({ annotation: body.annotation });
                    }
                    if (body.messages) {
                        this.setState({ messages: body.messages });
                    }
                }
                else if (err) {
                    this.setState({ messages: err });
                }
            });
        }
    }
    compilerCeId(i) {
        if (this.state.options[i].compiler.startsWith('clang'))
            return 'clang' + this.state.options[i].compiler.substr(6).replace('.', '') + '0';
        return 'g' + this.state.options[i].compiler.substr(4).replace('.', '');
    }
    optimCe(i) {
        switch (this.state.options[i].optim) {
            case 'G':
                return '-Og';
            case 'F':
                return '-Ofast';
            case 'S':
                return '-Os';
            default:
                return '-O' + this.state.options[i].optim;
        }
    }
    versionCe(i) {
        switch (this.state.options[i].cppVersion) {
            case '20':
                return '2a';
            case '17':
                return '1z';
            default:
                return this.state.options[i].cppVersion;
        }
    }
    b64UTFEncode(str) {
        return btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, function (match, v) {
            return String.fromCharCode(parseInt(v, 16));
        }));
    }
    decodeHash(str) {
        try {
            let base64ascii = str.substr(1);
            if (base64ascii) {
                return JSON.parse(atob(base64ascii));
            }
        } catch (err) {
            console.error(err);
        }
        return false;
    }
    optionsCe(i) {
        const cppVersion = '-std=c++' + this.versionCe(i);
        return cppVersion + ' ' + this.optimCe(i);
    }
    openCodeInCE() {
        let sessions = this.state.texts.map((t, i) => ({
            "id": i,
            "language": "c++",
            "source": t,
            "compilers": [{
                "id": this.compilerCeId(i),
                "options": this.optionsCe(i),
                "libs": [{
                    "name": "benchmark",
                    "ver": "140"
                }]
            }]
        }));
        var clientstate = {
            "sessions": sessions
        };
        var link = window.location.protocol + '//godbolt.org/clientstate/' + this.b64UTFEncode(JSON.stringify(clientstate));
        window.open(link, '_blank');
    }
    setDirty() {
        this.setState({
            clean: false,
            force: false
        });
    }
    codeChanged(code) {
        this.setState({
            texts: code
        });
        this.setDirty();
    }
    forceChanged(e) {
        this.setState({
            force: e.target.checked
        });
    }
    onOptionsChange(options) {
        this.setState({ options: options });
        this.setDirty();
    }
    onTitlesChange(titles) {
        this.setState({ titles: titles });
        this.setDirty();
    }
    toggleAnnotated(e) {
        this.setState({ isAnnotated: e.target.checked });
    }
    buttonHeight() {
        const run = document.getElementById('Run');
        if (run == null)
            return '5px';
        const compStyle = window.getComputedStyle(run, null);
        // We remove 4px more because for some reason otherwise it is possible that the CE button ends-up slightly bigger than the run button
        // Which because the whole toolbar is the same size, would start an infinit loop of 
        // "run" growing -> CE grows to react -> is bigger than run -> grows the toolbar
        return `calc(${compStyle.height} - ${compStyle.paddingTop} - ${compStyle.paddingBottom} - 4px)`;
    }
    closeTab(removedIndex) {
        let texts = this.state.texts;
        texts.splice(removedIndex, 1);
        let titles = this.state.titles;
        titles.splice(removedIndex, 1);
        let opts = this.state.options;
        opts.splice(removedIndex, 1);
        let messages = this.state.messages;
        this.setState({
            texts: texts,
            titles: titles,
            options: opts,
            messages: messages
        });

        this.setDirty();
    }
    addTab() {
        let texts = this.state.texts.concat(this.state.texts[this.state.index]);
        let titles = this.state.titles.concat(this.state.titles[this.state.index] + '2');
        let opts = this.state.options.concat({ ...this.state.options[this.state.index] });
        let messages = this.state.messages.concat('');
        this.setState({
            texts: texts,
            titles: titles,
            options: opts,
            messages: messages
        });

        this.setDirty();
    }
    render() {
        return (
            <Container fluid>
                <Row className="full-size">
                    <Col sm={6} className="full-size">
                        <div className="code-editor">
                            <WrappableTabs
                                titles={this.state.titles}
                                index={this.state.index}
                                setIndex={i => this.setState({ index: i })}
                                wrapped={this.state.textsWrapped}
                                changeWrapped={(w, c) => this.setState({ textsWrapped: w }, c)}
                                closeTab={(i) => this.closeTab(i)}
                                addTab={() => this.addTab()}
                                onTitlesChange={t => this.onTitlesChange(t)}
                                values={this.state.texts}
                                onChange={c => this.codeChanged(c)}
                                confirm
                                packed
                            >
                                <CodeEditor
                                    names={this.state.benchNames}
                                />
                            </WrappableTabs>
                        </div>
                    </Col>
                    <Col sm={6} className="flex-container">
                        <div className="fill-content">
                            <div className="fixed-content">
                                <Card body className="my-2">
                                    <WrappableTabs
                                        titles={this.state.titles}
                                        index={this.state.index}
                                        setIndex={i => this.setState({ index: i })}
                                        wrapped={this.state.optionsWrapped}
                                        changeWrapped={(w, c) => this.setState({ optionsWrapped: w }, c)}
                                        closeTab={(i) => this.closeTab(i)}
                                        addTab={() => this.addTab()}
                                        onTitlesChange={t => this.onTitlesChange(t)}
                                        values={this.state.options}
                                        onChange={c => this.onOptionsChange(c)}
                                        unwrapText="Configure Separately"
                                    >
                                        <CompileConfig />
                                    </WrappableTabs>
                                    <hr className="config-separator" />
                                    <ButtonToolbar className="justify-content-between">
                                        <Form inline>
                                            <Button variant="primary" onClick={() => this.sendCode()} disabled={this.state.sending} className="mr-2" id="Run"> <MdTimer /> Build Time</Button>
                                            {this.state.clean ? <FormCheck ref="force" type="checkbox" custom checked={this.state.force} id="clean-cache" onChange={this.forceChanged.bind(this)} label="Clear cached results" /> : null}
                                        </Form>
                                        <Form inline>
                                            <Button variant="outline-dark" onClick={() => this.openCodeInCE()} className="float-right"><img src="/ico/Compiler-Explorer.svg" style={{ height: this.buttonHeight() }} alt="Open in Compiler Explorer" /></Button>
                                        </Form>
                                    </ButtonToolbar>
                                    {this.state.sending ? <ProgressBar animated now={this.state.progress} /> : null}
                                </Card>
                            </div>
                            <Tab.Container defaultActiveKey="charts">
                                {this.state.graph.length > 0 ? (
                                    <Nav variant="tabs">
                                        <Nav.Item>
                                            <Nav.Link eventKey="charts">Charts</ Nav.Link>
                                        </Nav.Item>
                                        <Nav.Item>
                                            <Nav.Link eventKey="includes">Includes</ Nav.Link>
                                        </Nav.Item>
                                        <Nav.Item>
                                            <Nav.Link eventKey="asm">Assembly</ Nav.Link>
                                        </Nav.Item>
                                        <Nav.Item>
                                            <Nav.Link eventKey="pp">Preprocessed</ Nav.Link>
                                        </Nav.Item>
                                    </Nav>
                                ) : null}
                                <Tab.Content className="fill-content">
                                    <Tab.Pane eventKey="charts" className="fill-content">
                                        <TimeChart benchmarks={this.state.graph}
                                            id={this.state.location}
                                            chartIndex={this.state.chartIndex}
                                            onNamesChange={n => this.setState({ benchNames: n })}
                                            onDescriptionChange={d => this.props.onDescriptionChange(d)}
                                            specialPalette={this.props.specialPalette}
                                            dataChoices={chartData}
                                            changeDisplay={d => this.setState({ chartIndex: d })}
                                        />
                                    </Tab.Pane>
                                    <Tab.Pane eventKey="includes" className="fill-content" >
                                        <OutputTabs values={this.state.includes} index={this.state.index} setIndex={i => this.setState({ index: i })} titles={this.state.titles}>
                                            <IncludesDisplay />
                                        </OutputTabs>
                                    </Tab.Pane>
                                    <Tab.Pane eventKey="asm" className="fill-content">
                                        <OutputTabs values={this.state.asm} index={this.state.index} setIndex={i => this.setState({ index: i })} titles={this.state.titles}>
                                            <DisplayEditor language="asm" />
                                        </OutputTabs>
                                    </Tab.Pane>
                                    <Tab.Pane eventKey="pp" className="fill-content">
                                        <OutputTabs values={this.state.pp} index={this.state.index} setIndex={i => this.setState({ index: i })} titles={this.state.titles}>
                                            <DisplayEditor language="cpp" />
                                        </OutputTabs>
                                    </Tab.Pane>
                                </Tab.Content>
                            </Tab.Container>
                            <div className="fixed-content">
                                <OutputTabs values={this.state.messages} index={this.state.index} setIndex={i => this.setState({ index: i })} titles={this.state.titles} >
                                    <BashOutput />
                                </OutputTabs>
                            </div>
                        </div>
                    </Col>
                </Row>
            </Container>
        );
    }
}

export default Benchmark;
