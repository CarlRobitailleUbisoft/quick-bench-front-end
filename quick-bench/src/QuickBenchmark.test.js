import React from 'react';
import ReactDOM from 'react-dom';
import Benchmark from './QuickBenchmark';

import Fetch from "components/Fetch.js";

jest.mock("react-monaco-editor", () => () => <div></div>);

var searchedId;

jest.mock("components/Fetch.js", () => {
    return {
        __esModule: true,
        default: {
            fetchContent: jest.fn(),
            fetchResults: jest.fn()
        }
    }
});

it('doesnt load when there is no id', () => {
      Fetch.fetchContent.mockClear(); 
      const div = document.createElement('div');
    ReactDOM.render(<Benchmark/>, div);
    expect(Fetch.fetchContent.mock.calls.length).toBe(0);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('loads passed ids', () => {
      Fetch.fetchContent.mockClear(); 
    const div = document.createElement('div');
    ReactDOM.render(<Benchmark id={'abcd'}/>, div);
    expect(Fetch.fetchContent.mock.calls.length).toBe(1);
    expect(Fetch.fetchContent.mock.calls[0][1]).toBe('abcd');
    let callback = Fetch.fetchContent.mock.calls[0][2];
   
    ReactDOM.unmountComponentAtNode(div);
  });
