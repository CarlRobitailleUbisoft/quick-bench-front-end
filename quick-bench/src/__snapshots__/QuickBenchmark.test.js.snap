// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays id with annotations 1`] = `
<div
  className="fill-content container-fluid"
>
  <div
    className="fill-row row"
  >
    <div
      className="full-size col-sm-6"
    >
      <div
        className="code-editor"
      >
        <div
          className="flex-container"
        >
          <MonacoEditor
            editorDidMount={[Function]}
            language="cpp"
            onChange={[Function]}
            options={
              Object {
                "selectOnLineNumbers": true,
              }
            }
            value="static void StringCreation(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    std::string created_string(\\"hello\\");
    // Make sure the variable is not optimized away by compiler
    benchmark::DoNotOptimize(created_string);
  }
}
// Register the function as a benchmark
BENCHMARK(StringCreation);

static void StringCopy(benchmark::State& state) {
  // Code before the loop is not measured
  std::string x = \\"hello\\";
  for (auto _ : state) {
    std::string copy(x);
  }
}
BENCHMARK(StringCopy);
"
          />
        </div>
      </div>
    </div>
    <div
      className="flex-container col-sm-6"
    >
      <div
        className="fill-content"
      >
        <div
          className="fixed-content"
        >
          <div
            className="my-2 card"
          >
            <div
              className="card-body"
            >
              <div
                className="btn-toolbar"
                role="toolbar"
              >
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="compiler"
                    onClick={[Function]}
                    type="button"
                  >
                    compiler = Clang 10.0
                  </button>
                </div>
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="language"
                    onClick={[Function]}
                    type="button"
                  >
                    std = c++20
                  </button>
                </div>
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="optim"
                    onClick={[Function]}
                    type="button"
                  >
                    optim = O3
                  </button>
                </div>
                <div
                  className="dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="libc"
                    onClick={[Function]}
                    type="button"
                  >
                    STL = libstdc++(GNU)
                  </button>
                </div>
              </div>
              <hr
                className="config-separator"
              />
              <div
                className="justify-content-between btn-toolbar"
                role="toolbar"
              >
                <form
                  className="form-inline"
                >
                  <button
                    className="mr-2 btn btn-primary"
                    disabled={false}
                    id="Run"
                    onClick={[Function]}
                    type="button"
                  >
                     
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M15 1H9v2h6V1zm-4 13h2V8h-2v6zm8.03-6.61l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61zM12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                      />
                    </svg>
                     Run Benchmark
                  </button>
                  <div
                    className="mr-2 custom-control custom-checkbox"
                  >
                    <input
                      checked={true}
                      className="custom-control-input"
                      disabled={false}
                      id="disassembly"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="custom-control-label"
                      htmlFor="disassembly"
                      title=""
                    >
                      Record disassembly
                    </label>
                  </div>
                  <div
                    className="custom-control custom-checkbox"
                  >
                    <input
                      checked={false}
                      className="custom-control-input"
                      disabled={false}
                      id="clean-cache"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="custom-control-label"
                      htmlFor="clean-cache"
                      title=""
                    >
                      Clear cached results
                    </label>
                  </div>
                </form>
                <form
                  className="form-inline"
                >
                  <button
                    className="mr-2 btn btn-outline-dark"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    type="button"
                  >
                    <img
                      alt="Open in Compiler Explorer"
                      className="line-img"
                      src="Compiler-Explorer.svg"
                    />
                  </button>
                  <button
                    className="mr-2 btn btn-outline-dark"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    type="button"
                  >
                    <img
                      alt="Open in CPP Insights"
                      className="line-img"
                      src="cppinsights.svg"
                    />
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <div
          className="nav nav-tabs"
          onKeyDown={[Function]}
          role="tablist"
        >
          <div
            className="nav-item"
          >
            <a
              aria-selected={true}
              className="nav-link active"
              data-rb-event-key="charts"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="tab"
            >
              Charts
            </a>
          </div>
          <div
            className="nav-item"
          >
            <a
              aria-selected={false}
              className="nav-link"
              data-rb-event-key="asm"
              href="#"
              onClick={[Function]}
              onKeyDown={[Function]}
              role="tab"
              tabIndex={-1}
            >
              Assembly
            </a>
          </div>
        </div>
        <div
          className="fill-content tab-content"
        >
          <div
            aria-hidden={false}
            aria-labelledby={null}
            className="fade fill-content tab-pane active show"
            id={null}
            role="tabpanel"
          >
            <TimeChart
              afterTooltip={true}
              colors={
                Array [
                  Array [
                    "#5ed9cd",
                    "#f5b5c0",
                  ],
                ]
              }
              data={
                Array [
                  Array [
                    4.977504785737881,
                    14.969946162155678,
                  ],
                ]
              }
              dataLabels={
                Array [
                  "cpu_time",
                ]
              }
              fill={true}
              id="abcd"
              labels={
                Array [
                  "StringCreation",
                  "StringCopy",
                ]
              }
              legend={false}
              less="faster"
              more="slower"
              title={
                Array [
                  "ratio (CPU time / Noop time)",
                  "Lower is faster",
                ]
              }
              type="bar"
              xaxis="bar"
              yaxis="linear"
            >
              <div
                className="force-cb custom-control custom-checkbox custom-control-inline"
              >
                <input
                  checked={false}
                  className="custom-control-input"
                  disabled={false}
                  id="Noop"
                  onChange={[Function]}
                  type="checkbox"
                />
                <label
                  className="custom-control-label"
                  htmlFor="Noop"
                  title=""
                >
                  Show Noop bar
                </label>
              </div>
            </TimeChart>
          </div>
          <div
            aria-hidden={true}
            aria-labelledby={null}
            className="fade fill-content tab-pane"
            id={null}
            role="tabpanel"
          >
            <nav
              className="nav nav-tabs"
              onKeyDown={[Function]}
              role="tablist"
            />
            <div
              className="tab-content"
            />
            <div
              className="flex-container"
            >
              <MonacoEditor
                editorDidMount={[Function]}
                language="asm"
                options={
                  Object {
                    "lineNumbers": [Function],
                    "lineNumbersMinChars": 10,
                    "readOnly": true,
                    "selectOnLineNumbers": true,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`displays id without annotations 1`] = `
<div
  className="fill-content container-fluid"
>
  <div
    className="fill-row row"
  >
    <div
      className="full-size col-sm-6"
    >
      <div
        className="code-editor"
      >
        <div
          className="flex-container"
        >
          <MonacoEditor
            editorDidMount={[Function]}
            language="cpp"
            onChange={[Function]}
            options={
              Object {
                "selectOnLineNumbers": true,
              }
            }
            value="static void StringCreation(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    std::string created_string(\\"hello\\");
    // Make sure the variable is not optimized away by compiler
    benchmark::DoNotOptimize(created_string);
  }
}
// Register the function as a benchmark
BENCHMARK(StringCreation);

static void StringCopy(benchmark::State& state) {
  // Code before the loop is not measured
  std::string x = \\"hello\\";
  for (auto _ : state) {
    std::string copy(x);
  }
}
BENCHMARK(StringCopy);
"
          />
        </div>
      </div>
    </div>
    <div
      className="flex-container col-sm-6"
    >
      <div
        className="fill-content"
      >
        <div
          className="fixed-content"
        >
          <div
            className="my-2 card"
          >
            <div
              className="card-body"
            >
              <div
                className="btn-toolbar"
                role="toolbar"
              >
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="compiler"
                    onClick={[Function]}
                    type="button"
                  >
                    compiler = GCC 8.3
                  </button>
                </div>
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="language"
                    onClick={[Function]}
                    type="button"
                  >
                    std = c++17
                  </button>
                </div>
                <div
                  className="mr-2 dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={false}
                    id="optim"
                    onClick={[Function]}
                    type="button"
                  >
                    optim = O2
                  </button>
                </div>
                <div
                  className="dropdown"
                  onKeyDown={[Function]}
                >
                  <button
                    aria-expanded={false}
                    aria-haspopup={true}
                    className="dropdown-toggle btn btn-outline-dark"
                    disabled={true}
                    id="libc"
                    onClick={[Function]}
                    type="button"
                  >
                    STL = libstdc++(GNU)
                  </button>
                </div>
              </div>
              <hr
                className="config-separator"
              />
              <div
                className="justify-content-between btn-toolbar"
                role="toolbar"
              >
                <form
                  className="form-inline"
                >
                  <button
                    className="mr-2 btn btn-primary"
                    disabled={false}
                    id="Run"
                    onClick={[Function]}
                    type="button"
                  >
                     
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      strokeWidth="0"
                      style={
                        Object {
                          "color": undefined,
                        }
                      }
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M15 1H9v2h6V1zm-4 13h2V8h-2v6zm8.03-6.61l1.42-1.42c-.43-.51-.9-.99-1.41-1.41l-1.42 1.42C16.07 4.74 14.12 4 12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9 9-4.03 9-9c0-2.12-.74-4.07-1.97-5.61zM12 20c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                      />
                    </svg>
                     Run Benchmark
                  </button>
                  <div
                    className="mr-2 custom-control custom-checkbox"
                  >
                    <input
                      checked={false}
                      className="custom-control-input"
                      disabled={false}
                      id="disassembly"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="custom-control-label"
                      htmlFor="disassembly"
                      title=""
                    >
                      Record disassembly
                    </label>
                  </div>
                  <div
                    className="custom-control custom-checkbox"
                  >
                    <input
                      checked={false}
                      className="custom-control-input"
                      disabled={false}
                      id="clean-cache"
                      onChange={[Function]}
                      type="checkbox"
                    />
                    <label
                      className="custom-control-label"
                      htmlFor="clean-cache"
                      title=""
                    >
                      Clear cached results
                    </label>
                  </div>
                </form>
                <form
                  className="form-inline"
                >
                  <button
                    className="mr-2 btn btn-outline-dark"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    type="button"
                  >
                    <img
                      alt="Open in Compiler Explorer"
                      className="line-img"
                      src="Compiler-Explorer.svg"
                    />
                  </button>
                  <button
                    className="mr-2 btn btn-outline-dark"
                    disabled={false}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]}
                    type="button"
                  >
                    <img
                      alt="Open in CPP Insights"
                      className="line-img"
                      src="cppinsights.svg"
                    />
                  </button>
                </form>
              </div>
            </div>
          </div>
        </div>
        <div
          className="fill-content tab-content"
        >
          <div
            aria-hidden={false}
            aria-labelledby={null}
            className="fade fill-content tab-pane active show"
            id={null}
            role="tabpanel"
          >
            <TimeChart
              afterTooltip={true}
              colors={
                Array [
                  Array [
                    "#5ed9cd",
                    "#f5b5c0",
                  ],
                ]
              }
              data={
                Array [
                  Array [
                    4.977504785737881,
                    14.969946162155678,
                  ],
                ]
              }
              dataLabels={
                Array [
                  "cpu_time",
                ]
              }
              fill={true}
              id="abcd"
              labels={
                Array [
                  "StringCreation",
                  "StringCopy",
                ]
              }
              legend={false}
              less="faster"
              more="slower"
              title={
                Array [
                  "ratio (CPU time / Noop time)",
                  "Lower is faster",
                ]
              }
              type="bar"
              xaxis="bar"
              yaxis="linear"
            >
              <div
                className="force-cb custom-control custom-checkbox custom-control-inline"
              >
                <input
                  checked={false}
                  className="custom-control-input"
                  disabled={false}
                  id="Noop"
                  onChange={[Function]}
                  type="checkbox"
                />
                <label
                  className="custom-control-label"
                  htmlFor="Noop"
                  title=""
                >
                  Show Noop bar
                </label>
              </div>
            </TimeChart>
          </div>
          <div
            aria-hidden={true}
            aria-labelledby={null}
            className="fade fill-content tab-pane"
            id={null}
            role="tabpanel"
          >
            <nav
              className="nav nav-tabs"
              onKeyDown={[Function]}
              role="tablist"
            />
            <div
              className="tab-content"
            />
            <div
              className="flex-container"
            >
              <MonacoEditor
                editorDidMount={[Function]}
                language="asm"
                options={
                  Object {
                    "lineNumbers": [Function],
                    "lineNumbersMinChars": 10,
                    "readOnly": true,
                    "selectOnLineNumbers": true,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
